lib/dslash_policy.cuh

```diff
@@ -31,31 +32,35 @@ namespace quda
      * Arrays used for the dynamic scheduling.
      */
     struct DslashCommsPattern {
-      int gatherCompleted[Nstream];
-      int previousDir[Nstream];
-      int commsCompleted[Nstream];
-      int dslashCompleted[Nstream];
+      static constexpr int nDim = 4;
+      static constexpr int nDir = 2;
+      static constexpr int nStream = nDim * nDir + 1;
+
+      std::array<int, nStream> gatherCompleted;
+      std::array<int, nStream> previousDir;
+      std::array<int, nStream> commsCompleted;
+      std::array<int, nStream> dslashCompleted;
       int commDimTotal;
       int completeSum;
 
       inline DslashCommsPattern(const int commDim[], bool gdr_send = false) :
-          commsCompleted {},
-          dslashCompleted {},
-          completeSum(0)
+        commsCompleted {},
+        dslashCompleted {},
+        completeSum(0)
       {
 
-        for (int i = 0; i < Nstream - 1; i++) gatherCompleted[i] = gdr_send ? 1 : 0;
-        gatherCompleted[Nstream - 1] = 1;
-        commsCompleted[Nstream - 1] = 1;
-        dslashCompleted[Nstream - 1] = 1;
+        for (int i = 0; i < nStream - 1; i++) gatherCompleted[i] = gdr_send ? 1 : 0;
+        gatherCompleted[nStream - 1] = 1;
```

Comment:
`device::get_default_stream()`?

Reply:
Or `device::get_default_stream_idx()`?

Reply:
I'm not following this comment here.  Please explain 😄

Reply:
I think we should use `device::get_default_stream_idx()` (which maps onto `nStream - 1`) instead of having explicit `nStream - 1`

Reply:
My point more broadly is I don't think we should have `nStream - 1` baked in anywhere (except a target-specific directory)

Reply:
Ok, I get your point now.  Regardless, the `nStream` in dslash_policy.cuh is not an architecture specific value, it's rather  the number of concurrent work streams we have when applying the dslash in 4 dimensions, in 2 directions with an additional work stream (`nStream = 2 * 4 + 1`).  The fact that the last entry in the arrays corresponds to the default work stream outside of the dslash is now a historical coincidence.

It has been discussed that dslash_policy.cuh is long overdue for a rewrite, and we could make this distinction more clear in said rewrite.

Reply:
Sounds good---I'd appreciate a comment in the code along these lines for future reference, perhaps just one near the top of the file because I know this pattern is repeated multiple times.

Reply:
3eb6eafeb